metagraph/src/annotation/int_matrix/rank_extended/tuple_csc_matrix.hpp

```diff
@@ -0,0 +1,203 @@
+    }
+    return row_values;
+}
+
+template <class BaseMatrix, class Values, class Delims>
+inline std::vector<typename TupleCSCMatrix<BaseMatrix, Values, Delims>::RowValues>
+TupleCSCMatrix<BaseMatrix, Values, Delims>::get_row_values(const std::vector<Row> &rows) const {
+    const auto &column_ranks = binary_matrix_.get_column_ranks(rows);
+    std::vector<RowValues> row_values(rows.size());
+    // TODO: reshape?
+    for (size_t i = 0; i < rows.size(); ++i) {
+        row_values[i].reserve(column_ranks[i].size());
+        for (auto [j, r] : column_ranks[i]) {
+            assert(r >= 1 && "matches can't have zero-rank");
+            size_t tuple_size = delimiters_[j].select1(r + 1) - delimiters_[j].select1(r) - 1;
+            row_values[i].emplace_back(j, tuple_size);
+        }
+    }
+    return row_values;
+}
+
+template <class BaseMatrix, class Values, class Delims>
+uint64_t TupleCSCMatrix<BaseMatrix, Values, Delims>::num_attributes() const {
+    uint64_t num_attributes = 0;
+    for (size_t j = 0; j < column_values_.size(); ++j) {
+        num_attributes += column_values_[j].size();
+    }
+    return num_attributes;
+}
+
+template <class BaseMatrix, class Values, class Delims>
+inline typename TupleCSCMatrix<BaseMatrix, Values, Delims>::RowTuples
+TupleCSCMatrix<BaseMatrix, Values, Delims>::get_row_tuples(Row row) const {
+    const auto &column_ranks = binary_matrix_.get_column_ranks(row);
+    RowTuples row_tuples;
+    row_tuples.reserve(column_ranks.size());
+    for (auto [j, r] : column_ranks) {
+        assert(r >= 1 && "matches can't have zero-rank");
+        size_t begin = delimiters_[j].select1(r) + 1 - r;
+        size_t end = delimiters_[j].select1(r + 1) - r;
+        Tuple tuple;
+        tuple.reserve(end - begin);
+        for (size_t t = begin; t < end; ++t) {
+            tuple.push_back(column_values_[j][t]);
+        }
+        row_tuples.emplace_back(j, std::move(tuple));
+    }
+    return row_tuples;
+}
+
+template <class BaseMatrix, class Values, class Delims>
+inline std::vector<typename TupleCSCMatrix<BaseMatrix, Values, Delims>::RowTuples>
+TupleCSCMatrix<BaseMatrix, Values, Delims>::get_row_tuples(const std::vector<Row> &rows) const {
+    const auto &column_ranks = binary_matrix_.get_column_ranks(rows);
+    std::vector<RowTuples> row_tuples(rows.size());
+    // TODO: reshape?
+    for (size_t i = 0; i < rows.size(); ++i) {
+        row_tuples[i].reserve(column_ranks[i].size());
+        for (auto [j, r] : column_ranks[i]) {
+            assert(r >= 1 && "matches can't have zero-rank");
+            size_t begin = delimiters_[j].select1(r) + 1 - r;
+            size_t end = delimiters_[j].select1(r + 1) - r;
+            Tuple tuple;
+            tuple.reserve(end - begin);
+            for (size_t t = begin; t < end; ++t) {
+                tuple.push_back(column_values_[j][t]);
+            }
+            row_tuples[i].emplace_back(j, std::move(tuple));
+        }
```

Comment:
I agree, I think it might be more cache efficient to first aggregate a list of all accessed column IDs, then iterate through each column once to populate this vector.